google/cloud/bigquery/dtypes.py

```diff
@@ -0,0 +1,435 @@
+        self, *, axis: Optional[int] = None, skipna: bool = True, **kwargs
+    ) -> Scalar:
+        pandas.compat.numpy.function.validate_max((), kwargs)
+        result = pandas.core.nanops.nanmax(
+            values=self._ndarray, axis=axis, mask=self.isna(), skipna=skipna
+        )
+        return self._box_func(result)
+
+    if pandas_release >= (1, 2):
+
+        def median(
+            self,
+            *,
+            axis: Optional[int] = None,
+            out=None,
+            overwrite_input: bool = False,
+            keepdims: bool = False,
+            skipna: bool = True,
+        ):
+            pandas.compat.numpy.function.validate_median(
+                (),
+                {"out": out, "overwrite_input": overwrite_input, "keepdims": keepdims},
+            )
+            result = pandas.core.nanops.nanmedian(
+                self._ndarray, axis=axis, skipna=skipna
+            )
+            return self._box_func(result)
+
+
+@pandas.core.dtypes.dtypes.register_extension_dtype
+class TimeDtype(_BaseDtype):
+    """
+    Extension dtype for time data.
+    """
+
+    name = "time"
+    type = datetime.time
+
+    def construct_array_type(self):
+        return TimeArray
+
+
+class TimeArray(_BaseArray):
+    """
+    Pandas array type containing time data
+    """
+
+    # Data are stored as datetime64 values with a date of Jan 1, 1970
+
+    dtype = TimeDtype()
+    _epoch = datetime.datetime(1970, 1, 1)
+    _npepoch = numpy.datetime64(_epoch)
+
+    @classmethod
+    def _datetime(cls, scalar):
+        if isinstance(scalar, datetime.time):
+            return datetime.datetime.combine(cls._epoch, scalar)
+        elif isinstance(scalar, str):
+            # iso string
+            h, m, s = map(float, scalar.split(":"))
+            s, us = divmod(s, 1)
+            return datetime.datetime(
+                1970, 1, 1, int(h), int(m), int(s), int(us * 1000000)
+            )
+        else:
+            raise TypeError("Invalid value type", scalar)
+
+    def _box_func(self, x):
+        if pandas.isnull(x):
+            return None
+
+        try:
+            return x.astype("<M8[us]").astype(datetime.datetime).time()
+        except AttributeError:
+            x = numpy.datetime64(x)
+            return x.astype("<M8[us]").astype(datetime.datetime).time()
+
+    __return_deltas = {"timedelta", "timedelta64", "timedelta64[ns]", "<m8", "<m8[ns]"}
+
+    def astype(self, dtype, copy=True):
+        deltas = self._ndarray - self._npepoch
+        stype = str(dtype)
+        if stype in self.__return_deltas:
+            return deltas
+        elif stype.startswith("timedelta64[") or stype.startswith("<m8["):
+            return deltas.astype(dtype, copy=False)
+        else:
+            return super().astype(dtype, copy=copy)
+
+
+@pandas.core.dtypes.dtypes.register_extension_dtype
+class DateDtype(_BaseDtype):
+    """
+    Extension dtype for time data.
+    """
+
+    name = "date"
+    type = datetime.date
```

Comment:
do we define this or does it come from pandas? If we do, we usually add a `_` suffix to anything that has the same name as a built-in function. `type_`.

Reply:
From pandas
https://pandas.pydata.org/docs/reference/api/pandas.api.extensions.ExtensionDtype.html